firebase-installations/src/androidTest/java/com/google/firebase/installations/FirebaseInstallationsInstrumentedTest.java

```diff
@@ -26,4 +47,99 @@
+    when(backendClientReturnsOk.createFirebaseInstallation(
+            anyString(), anyString(), anyString(), anyString()))
+        .thenReturn(
+            InstallationResponse.builder()
+                .setName("/projects/123456789/installations/fid")
+                .setRefreshToken("refresh-token")
+                .setAuthToken(
+                    InstallationTokenResult.builder()
+                        .setToken("auth-token")
+                        .setTokenExpirationTimestampMillis(1000L)
+                        .build())
+                .build());
+    when(backendClientReturnsError.createFirebaseInstallation(
+            anyString(), anyString(), anyString(), anyString()))
+        .thenThrow(
+            new FirebaseInstallationServiceException(
+                "SDK Error", FirebaseInstallationServiceException.Status.SERVER_ERROR));
+    when(cacheReturnsError.insertOrUpdateCacheEntry(any())).thenReturn(false);
+    when(cacheReturnsError.readCacheEntryValue()).thenReturn(null);
+  }
+
+  @After
+  public void cleanUp() throws Exception {
+    actualCache.clear();
+  }
+
+  @Test
+  public void testCreateFirebaseInstallation_CacheOk_BackendOk() throws Exception {
+    FirebaseInstallations firebaseInstallations =
+        new FirebaseInstallations(firebaseApp, actualCache, backendClientReturnsOk);
+
+    // No exception, means success.
+    assertThat(Tasks.await(firebaseInstallations.getId())).isNotEmpty();
+    FiidCacheEntryValue entryValue = actualCache.readCacheEntryValue();
+    assertThat(entryValue.getFirebaseInstallationId()).isNotEmpty();
+    assertThat(entryValue.getCacheStatus()).isEqualTo(FiidCache.CacheStatus.REGISTERED);
+  }
+
+  @Test
+  public void testCreateFirebaseInstallation_CacheOk_BackendError() throws Exception {
+    FirebaseInstallations firebaseInstallations =
+        new FirebaseInstallations(firebaseApp, actualCache, backendClientReturnsError);
+
+    // Expect exception
+    try {
+      Tasks.await(firebaseInstallations.getId());
+      fail();
+    } catch (ExecutionException expected) {
+      Throwable cause = expected.getCause();
+      assertThat(cause).isInstanceOf(FirebaseInstallationsException.class);
+      assertThat(((FirebaseInstallationsException) cause).getStatus())
+          .isEqualTo(FirebaseInstallationsException.Status.SDK_INTERNAL_ERROR);
+    }
+
+    FiidCacheEntryValue entryValue = actualCache.readCacheEntryValue();
+    assertThat(entryValue.getFirebaseInstallationId()).isNotEmpty();
+    assertThat(entryValue.getCacheStatus()).isEqualTo(FiidCache.CacheStatus.REGISTER_ERROR);
+  }
+
+  @Test
+  public void testCreateFirebaseInstallation_CacheError_BackendOk() throws InterruptedException {
+    FirebaseInstallations firebaseInstallations =
+        new FirebaseInstallations(firebaseApp, cacheReturnsError, backendClientReturnsOk);
+
+    // Expect exception
+    try {
+      Tasks.await(firebaseInstallations.getId());
```

Comment:
out of curiosity:
Does this framework not have "assertThrows"?

Reply:
There isnt one. However, I found RC using one by adding Asssert in Test Utils.

https://github.com/firebase/firebase-android-sdk/blob/5301121df4727b357293b3b7de00555362f90170/firebase-config/src/test/java/com/google/firebase/remoteconfig/testutil/Assert.java

Rest of the firebase-android-sdk teams have handled the same way as me.

Reply:
it was only added in junit5, and this build is using junit4. https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html#assertThrows-java.lang.Class-org.junit.jupiter.api.function.Executable-